PWGHF/TableProducer/treeCreatorDplusToPiKPi.cxx

```diff
@@ -0,0 +1,429 @@
+                  collision::PosZ,
+                  full::IsEventReject,
+                  full::RunNumber);
+
+DECLARE_SOA_TABLE(HfCand3ProngFullParticles, "AOD", "HFCAND3PFullP",
+                  collision::BCId,
+                  full::Pt,
+                  full::Eta,
+                  full::Phi,
+                  full::Y,
+                  hf_cand_3prong::FlagMcMatchRec,
+                  hf_cand_3prong::OriginMcGen);
+} // namespace o2::aod
+
+/// Writes the full information in an output TTree
+struct HfTreeCreatorDplusToPiKPi {
+  Produces<o2::aod::HfCand3ProngFull> rowCandidateFull;
+  Produces<o2::aod::HfCand3ProngFullEvents> rowCandidateFullEvents;
+  Produces<o2::aod::HfCand3ProngFullParticles> rowCandidateFullParticles;
+  Produces<o2::aod::HfCand3ProngLite> rowCandidateLite;
+
+  Configurable<int> selectionFlagDplus{"selectionFlagDplus", 1, "Selection Flag for Dplus"};
+  Configurable<bool> fillCandidateLiteTable{"fillCandidateLiteTable", false, "Switch to fill lite table with candidate properties"};
+
+  Filter filterSelectCandidates = aod::hf_sel_candidate_dplus::isSelDplusToPiKPi >= selectionFlagDplus;
+
+  void init(InitContext const&)
+  {
+  }
+
+  template <typename T>
+  void fillEvent(const T& collision, int isEventReject, int runNumber)
+  {
+    rowCandidateFullEvents(
+      collision.bcId(),
+      collision.numContrib(),
+      collision.posX(),
+      collision.posY(),
+      collision.posZ(),
+      isEventReject,
+      runNumber);
+  }
+
+  template <typename T, typename U>
+  void fillLiteTable(const T& candidate, const U& prong0, const U& prong1, const U& prong2,
+                     int selectionFlag, double invMass, double y, int8_t flagMc, int8_t origin)
+  {
+    rowCandidateLite(
+      candidate.chi2PCA(),
+      candidate.decayLength(),
+      candidate.decayLengthXY(),
+      candidate.decayLengthNormalised(),
+      candidate.decayLengthXYNormalised(),
+      candidate.ptProng0(),
+      candidate.ptProng1(),
+      candidate.ptProng2(),
+      candidate.impactParameter0(),
+      candidate.impactParameter1(),
+      candidate.impactParameter2(),
+      prong0.tpcNSigmaPi(),
+      prong0.tpcNSigmaKa(),
+      prong0.tofNSigmaPi(),
+      prong0.tofNSigmaKa(),
+      prong1.tpcNSigmaPi(),
+      prong1.tpcNSigmaKa(),
+      prong1.tofNSigmaPi(),
+      prong1.tofNSigmaKa(),
+      prong2.tpcNSigmaPi(),
+      prong2.tpcNSigmaKa(),
+      prong2.tofNSigmaPi(),
+      prong2.tofNSigmaKa(),
+      selectionFlag,
+      invMass,
+      candidate.pt(),
+      candidate.cpa(),
+      candidate.cpaXY(),
+      candidate.maxNormalisedDeltaIP(),
+      candidate.eta(),
+      candidate.phi(),
+      y,
+      flagMc,
+      origin);
+  }
+
+  template <typename T, typename U>
+  void fillTable(const T& candidate, const U& prong0, const U& prong1, const U& prong2, int selectionFlag, double invMass,
+                 double ct, double y, double e, int8_t flagMc, int8_t origin)
```

Comment:
Are you passing all these arguments from outside to avoid code duplication inside `fillTable` and `fillLiteTable`?

Reply:
I mean that all the arguments you pass can be accessed from `candidate`.

Reply:
But now I see that there is also a big overlap between the two `process` functions which could also be factored out into a single filling function.

Reply:
> > I mean that all the arguments you pass can be accessed from `candidate`.
> 
> Indeed, some of them can be. However, for `prong0,1,2`, it does not compile if I define them inside the `fillTable` function.

OK, fair enough. Let's keep the prongs where they are then. :-)